### PR TITLE
Use `line-trim-offset` to replace the line gradient stops for route line update.

### PR DESCRIPTION
**Pull request by ShanMa1991**
### Description
This Pr is to support using the `lineLayer.line-trim-offset` to update the route line when `routeLineTracksTraversal` is enabled during active navigation. 

### Implementation
- [x] Use `lineLayer.line-trim-offset` to trim the route line layer instead of updating the gradient stops.
- [x] Remove the stored gradient stops for route line and restricted area in `NavigationMapView`.
- [x] Support the instant route line change when `showsRestrictedAreasOnRoute`, `crossfadesCongestionSegments` and `routeLineTracksTraversal` property change during active navigation session.
- [x] Fix the test.

### Screenshots or Gifs
in `Example` App, switch the route line property during active navigation with `Resume` button.

switch `navigationMapView.crossfadesCongestionSegments` when `routeLineTracksTraversal` enabled.
crossfadesCongestionSegments

switch `navigationMapView.showsRestrictedAreasOnRoute` when `routeLineTracksTraversal` enabled with mock restricted area.
showsRestrictedAreasOnRoute

switch `routeLineTracksTraversal` in active navigation
routeLineTracksTraversal

### CODE REVIEW

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 1]
```diff
@@ -321,13 +330,24 @@ class VanishingRouteLineTests: TestCase {
...
         // When different congestion levels have same color, the line gradient stops are expected to combine these congestion level.
+        navigationMapView.trafficUnknownColor = UIColor.blue
         navigationMapView.trafficModerateColor = navigationMapView.trafficUnknownColor
         navigationMapView.routes = [route]
         navigationMapView.routeLineTracksTraversal = true
         navigationMapView.show([route], legIndex: 0)
         
-        let expectedGradientStops = [0.0 : navigationMapView.trafficUnknownColor]
-        XCTAssertEqual(expectedGradientStops, navigationMapView.currentLineGradientStops, "Failed to combine the same color of congestion segment.")
+        let expectedExpressionString = "[step, [line-progress], [rgba, 0.0, 0.0, 255.0, 1.0], 0.0, [rgba, 0.0, 0.0, 255.0, 1.0]]"
```
[the remark this answers is not in the document]
Comment (ShanMa1991) [reply Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:340]: Thanks for the suggestion, the comments are added to provide the clarification of the line gradients.

File: Sources/MapboxNavigation/NavigationMapView.swift [hunk 2]
```diff
@@ -596,10 +550,7 @@ open class NavigationMapView: UIView {
         pendingCoordinateForRouteLine = nil
     }
     
-    @discardableResult func addRouteRestrictedAreaLayer(_ route: Route,
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/NavigationMapView.swift:599]: I think that it would've been better to land fix for restricted areas position in different PR. It's hard to follow what was exactly wrong and how such issue was fixed.

File: Sources/MapboxNavigation/NavigationMapView.swift [hunk 3]
```diff
@@ -314,16 +302,19 @@ open class NavigationMapView: UIView {
             
             let offset = (route.distance - routeAlternative.infoFromDeviationPoint.distance) / route.distance
             parentLayerIdentifier = addRouteLayer(route,
-                                                  fractionTraveled: offset,
                                                   below: parentLayerIdentifier,
-                                                  reuseExistingLayer: true,
                                                   isMainRoute: false,
                                                   legIndex: nil)
+            if let altertiveRouteLayerIdentifier = parentLayerIdentifier {
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/NavigationMapView.swift:308]: Did you mean `alternativeRouteLayerIdentifier` in this case?

File: Sources/MapboxNavigation/NavigationMapView.swift (hunk 2 above)
Comment (ShanMa1991) [reply Sources/MapboxNavigation/NavigationMapView.swift:599]: The restricted area position is caused by the line gradient stops generation. When we remove the fractionTraveled and adjust the base color for empty restricted feature gradient stops generation, the issue will be fixed. It's hard to split them to two.

File: Sources/MapboxNavigation/NavigationMapView.swift (hunk 3 above)
Comment (ShanMa1991) [reply Sources/MapboxNavigation/NavigationMapView.swift:308]: Thanks for the suggestion, I'll update the name in https://github.com/mapbox/mapbox-navigation-ios/pull/4015

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 4]
```diff
@@ -321,13 +330,25 @@ class VanishingRouteLineTests: TestCase {
...
         // When different congestion levels have same color, the line gradient stops are expected to combine these congestion level.
+        navigationMapView.trafficUnknownColor = UIColor.blue
         navigationMapView.trafficModerateColor = navigationMapView.trafficUnknownColor
         navigationMapView.routes = [route]
         navigationMapView.routeLineTracksTraversal = true
         navigationMapView.show([route], legIndex: 0)
         
-        let expectedGradientStops = [0.0 : navigationMapView.trafficUnknownColor]
-        XCTAssertEqual(expectedGradientStops, navigationMapView.currentLineGradientStops, "Failed to combine the same color of congestion segment.")
+        // The moderate and unknow traffic congestion share the same color, so the line gradient is expected to combine same feature color into one stop.
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:339]: Small typo here as well: `unknown` instead of `unknow`.

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 5]
```diff
@@ -346,10 +367,14 @@ class VanishingRouteLineTests: TestCase {
...
         navigationMapView.travelAlongRouteLine(to: coordinate)
 
-        let fractionTraveled = navigationMapView.fractionTraveled
-        let fractionTraveledNextDown = Double(CGFloat(fractionTraveled).nextDown)
+        let expectedFractionTraveled = 0.3240769449298392
+        var actualFractionTraveled = navigationMapView.fractionTraveled
+        XCTAssertEqual(actualFractionTraveled, expectedFractionTraveled, accuracy: accuracyThreshold, "Failed to update route line when routeLineTracksTraversal enabled.")
         
-        var expectedExpressionString = "[step, [line-progress], [rgba, 0.0, 0.0, 255.0, 1.0], 0.0, [rgba, 0.0, 0.0, 0.0, 0.0], \(fractionTraveledNextDown), [rgba, 0.0, 0.0, 0.0, 0.0], \(fractionTraveled), [rgba, 0.0, 0.0, 255.0, 1.0], 0.9425498181625797, [rgba, 0.0, 0.0, 255.0, 1.0], 0.9425498181625799, [rgba, 255.0, 0.0, 0.0, 1.0]]"
+        // The line gradient is expected to have two parts. The first is from 0.0 to unknow traffic congestion feature end with trafficUnknownColor.
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:374]: Same here.

File: CHANGELOG.md [hunk 6]
```diff
@@ -25,6 +25,7 @@
 
 * Fixed the unmatched language issue of some localized strings in the feedback view. ([#4007](https://github.com/mapbox/mapbox-navigation-ios/pull/4007))
 * Added `NavigationSettings.utilizeSensorData` toggle to allow navigator to use additional device sensors data for better positioning. ([#3973](https://github.com/mapbox/mapbox-navigation-ios/pull/3973))
+* Fixed the issue of restricted layer over the whole route after the removal of continuous alternative route with a small portion of restricted road class at start. ([#4009](https://github.com/mapbox/mapbox-navigation-ios/pull/4009))
```
Comment (MaximAlien) [inline CHANGELOG.md:28]: To catch such issues in future we can consider adding new snapshot tests to `MapboxNavigationTests`.

Comment (ShanMa1991) [reply CHANGELOG.md:28]: Thanks for the suggestion, I'll open a ticket to track this issue.
